### PR TITLE
[DROOLS-5324] refactor building of non-native resources from the exec…

### PR DESCRIPTION
**Pull request by mariofusco**
…utable model

### CODE REVIEW

File: kie-api/src/main/java/org/kie/api/io/ResourceType.java [hunk 1]
```diff
@@ -30,24 +30,28 @@
...
-    private String defaultExtension;
+    private final String defaultExtension;
 
-    private String[] otherExtensions;
+    private final String[] otherExtensions;
 
-    private String defaultPath;
+    private final String defaultPath;
+
+    private final boolean isNative;
```
Comment (danielezonca) [inline kie-api/src/main/java/org/kie/api/io/ResourceType.java:43]: Can you please clarify what `native` means in this context?
Why DMN, FEEL and PMML are the only type not native? What about `JAVA`, `BPMN` and similar?

Minor note: it should be better to remove `is` prefix because otherwise the getter should be named `isIsNative` to be compliant with the naming conventions :)

File: kie-api/src/main/java/org/kie/api/io/ResourceType.java [hunk 2]
```diff
@@ -60,12 +64,25 @@ public static ResourceType addResourceTypeToRegistry(final String resourceType,
...
+
         ResourceType resource = new ResourceType( resourceType,
+                                                  isNative,
                                                   description,
                                                   defaultPath,
                                                   defaultExtension,
                                                   otherExtensions );
+
         CACHE.put( resourceType, resource );
+        resource.getAllExtensions().forEach( ext -> CACHE.put( "." + ext, resource ) );
```
Comment (danielezonca) [inline kie-api/src/main/java/org/kie/api/io/ResourceType.java:85]: Is this change related with the PR or it is just an unrelated fix?

File: kie-api/src/main/java/org/kie/api/io/ResourceType.java [hunk 3]
```diff
@@ -161,6 +178,7 @@ public static ResourceType addResourceTypeToRegistry(final String resourceType,
 
     /** PMML */
     public static final ResourceType PMML = addResourceTypeToRegistry("PMML",
+                                                                      false,
                                                                       "Predictive Model Markup Language",
                                                                       "src/main/resources",
                                                                       "pmml");
```
Comment (danielezonca) [inline kie-api/src/main/java/org/kie/api/io/ResourceType.java:184]: Are you sure this change has no impact on existing `kie-pmml` impl?

File: kie-api/src/main/java/org/kie/api/io/ResourceType.java [hunk 4]
```diff
@@ -248,12 +268,8 @@ public static ResourceType getResourceType(final String resourceType) {
...
 
     public static ResourceType determineResourceType(final String resourceName) {
-        for (ResourceType type : CACHE.values()) {
-            if (type.matchesExtension(resourceName)) {
-                return type;
-            }
-        }
-        return null;
+        int dotPos = resourceName.lastIndexOf( '.' );
+        return dotPos < 0 ? null : CACHE.get( resourceName.substring( dotPos ) );
```
Comment (danielezonca) [inline kie-api/src/main/java/org/kie/api/io/ResourceType.java:272]: This change can introduce a bug if (or when) we have a resource type extension that contains `.` (i.e. `gwt.xml`). I understand a `get` on a map is much more efficient than the previous version but I think it is a premature optimization.

File: kie-api/src/main/java/org/kie/api/io/ResourceType.java (hunk 1 above)
Comment (mariofusco) [reply kie-api/src/main/java/org/kie/api/io/ResourceType.java:43]: In this context `native` means something that it is build natively by drools-compiler without going through an assembler. Also note that `native` is a Java keyword and that's why I named the field `isNative`. I'm not entirely happy with the name, so any suggestion for a better one is very welcome.

File: kie-api/src/main/java/org/kie/api/io/ResourceType.java (hunk 2 above)
Comment (mariofusco) [reply kie-api/src/main/java/org/kie/api/io/ResourceType.java:85]: Not related and not a fix, just a minor performance improvement to avoid that it loops all the available resource types when trying to determine the type of a file.

Comment (danielezonca) [reply kie-api/src/main/java/org/kie/api/io/ResourceType.java:85]: Ok thanks for the clarification :+1:

File: kie-api/src/main/java/org/kie/api/io/ResourceType.java (hunk 4 above)
Comment (mariofusco) [reply kie-api/src/main/java/org/kie/api/io/ResourceType.java:272]: Functionally the public API of this class behaves exactly as before, I made just a perf improvement. If this breaks something (I don't think so) I'm hoping we will realized this with tests.

Comment (danielezonca) [reply kie-api/src/main/java/org/kie/api/io/ResourceType.java:272]: It doesn't break now because we have no `ResourceType` that has a `.` as part of the extension.
This code assumes a constraint on extension format (aka split on ``lastIndexOf `.`)`` to get extension) that is implicit and not enforced.
The method used in the previous code ([matchesExtension](https://github.com/kiegroup/droolsjbpm-knowledge/blob/bc2f3aae35e4d7a419abe8afa2751cd931587118/kie-api/src/main/java/org/kie/api/io/ResourceType.java#L278)) doesn't suffer of the same issue because the used condition is `endsWith("." + defaultExtension)`

In short, this type is legit and breaks the proposed code
```java
public static final ResourceType BROKEN = addResourceTypeToRegistry("BROKEN",
                                                                    "Broken resrouce type",
                                                                    "src/main/resources",
                                                                    "broken.extension");
```

File: kie-api/src/main/java/org/kie/api/io/ResourceType.java [hunk 5]
```diff
@@ -30,24 +30,28 @@
...
 
-    private String defaultPath;
+    private final String defaultPath;
+
+    private final boolean fullyCoveredByExecModel;
 
     private static final Map<String, ResourceType> CACHE = Collections.synchronizedMap(new HashMap<String, ResourceType>());
 
     public ResourceType(String name,
+                        boolean fullyCoveredByExecModel,
```
Comment (gitgabrio) [inline kie-api/src/main/java/org/kie/api/io/ResourceType.java:48]: @mariofusco 
Looking at how this is used inside the other PR (https://github.com/kiegroup/drools/pull/2926) I think this is not needed and misleading
1) it must have always the same value for a given type
2) there is always a direct relationship 1:1 between the type and this flag
3) it seems to only add level of indirectness (inside CanonicalKieModule.NON_MODEL_RESOURCES)
